okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java

```diff
@@ -344,6 +347,73 @@ static Status toGrpcStatus(ErrorCode code) {
     return ERROR_CODE_TO_STATUS.get(code);
   }
 
+  private Socket upgradeToTls(Socket socket, SSLSocketFactory sslSocketFactory) throws IOException {
+    SSLSocket sslSocket = (SSLSocket) sslSocketFactory.createSocket(
+        socket, authorityHost, address.getPort(), true /* auto close */);
+    final String[] supportedTlsProtocols = sslSocket.getSupportedProtocols();
+    Set<String> supportedTlsProtocolsSet = new HashSet<String>(supportedTlsProtocols.length);
+    for (int i = 0; i < supportedTlsProtocols.length; ++i) {
+      supportedTlsProtocolsSet.add(supportedTlsProtocols[i]);
+    }
+    List<String> enableTlsProtocols = new ArrayList<String>();
+    addIfSupported(supportedTlsProtocolsSet, enableTlsProtocols,
+        "TLSv1.2", "TLSv1.1", "TLSv1");
+    sslSocket.setEnabledProtocols(
+        enableTlsProtocols.toArray(new String[enableTlsProtocols.size()]));
+
+    final String[] supportedCiphers = sslSocket.getSupportedCipherSuites();
+    Set<String> supportedCiphersSet = new HashSet<String>(supportedCiphers.length);
```

Comment:
FYI, using supportedCiphers.length still causes the HashSet to resize, as it does not take into account the loadFactor. The "capacity" of the hash table is the number slots. The actual capacity can be greater or smaller than that based on the loadFactor.

Because the default loadFactor is .75, when >75% of supportedCiphers.length is added the table will resize.

This makes properly sizing a hash table more annoying. Something more like:
```java
new HashSet<String>((int) (supportedCiphers.length / 0.75) + 1);
// or
new HashSet<String>(supportedCiphers.length * 4 / 3 + 1);
```

That is a pain and very few people would understand the reason for the madness, so sometimes I will just do ```2 * supportedCiphers.length``` or just give up and stop caring.